### DISCUSSION TITLE
Unable to get all images from a page

### DISCUSSION THREAD

**Discussion by mahdyshabeeb** [pymupdf/PyMuPDF#2714] [Q&A] [answered]
When using the `page.get_image()` method, only some of the images on the given page are returned. Is there any parameter I need to add to be able to get all the images? Or is there a different method I can use?

**Comment by JorjMcKie** [comment]
Atypical "Discussions" post. Let me convert first.

**Comment by JorjMcKie** [comment, marked as answer]
This may have several reasons:

* There exist "inline" images. They have no xref and exist only inside the page appearace source (`/Contents` objects). You can still extract them via `page.get_text("dict")` (or "rawdict").
* There exist vector graphics. The more complex ones may look like images, but they aren't and they also have no xref. You can extract those via `page.get_drawings()`.
* Some PDF creators make annotations with an image instead of a fill color (e.g. buttons). They do not appear in the `page.get_images()` although they do have an xref. We don't yet support to locate them.

**Reply by mahdyshabeeb** [reply]
You are right. The images I was missing are vector graphics that's why they were not returned.

However, for some reason, the method `page.get_text("dict")` does not return all the images returned by `page.get_images()`. On the other hand, using "html" instead of "dict" returns all the images.

Now I have 3 questions (please let me know if I should open a new discussion):

1. Is it possible to convert these drawings to an image/ images? I have applied [the code snippit provided in the documentation](https://pymupdf.readthedocs.io/en/latest/document.html#Document.get_page_images) (with minor modifications) which made a new pdf file out of them but my question is if I can decode them to get their binary representation (which can for example be displayed by `PIL.Image.open()`)
2. Is there a known reason why `page.get_text("dict")` does not return all images?
3. Is there a way to get the positions of the images using the `page.get_images()` method or should I use page.get_text() instead?

**Comment by JorjMcKie** [comment]
Your questions:

> However, for some reason, the method page.get_text("dict") does not return all the images returned by page.get_images(). On the other hand, using "html" instead of "dict" returns all the images.

This sounds weird a bit. `get_images()` (same as `doc.get_page_images()`) is just a report for what the PDF object definition of the page has to say. Depending on how the PDF was made, this list may sometimes reference all images of the full PDF - not only the ones, that this page displays.
There is a function, `page.clean_contents()`, which cleans up such situations.
`page.get_image_info()` reports meta information of each image actually displayed (xref-ed or not). Should be the same as the respective subset of `page.get_text("dict")`.
If html output returns more than these, then it may only be caused by images occurring in annotations as described above.
Otherwise please send me the example.

Image positions on the page:
Use `page.get_image_rects(xref)`. This returns the bbox plus optionally a matrix - for each occurrence of the same image on that page (which does happen). So this is a list.
The optional matrix contains information about what was happening to the original image to fit it into the bbox: scaling, rotation, flipping, etc. Please see documentation for details.

Recovering (extracting) Images:
There are several options:
1. Take the binary returned by `get_text("[raw]dict")`. Please note that unfortunately there is no way to access transparency here - or even knowing whether the original image **_was_** transparent. But independent from that you also get the appropriate file extension (like "png"). You can save the image via `pathlib.Path(f"image.{img['ext']}").write_bytes(img["image"])`. Or open it as a PIL Image: define a file-like object first using `fp = io.BytesIO(img["image"])`, then `pil_img = PIL.Image.open(fp)`. Or make a `fitz.Pixmap` from it: `pix = fitz.Pixmap(img["image"])`.
2. If you have the image xref, you can do `img = doc.extract_image(xref)` which returns a dictionary with image meta information and the binary. Process this similar to as before. Here, you do have access to transparency information: there is an "smask" key in the dictionary. If this is positive, then this is the xref of the transparency object (PDF separates this unfortunately). Please see the documentation of the method to learn how the full original image can be recovered in this case.

**Reply by mahdyshabeeb** [reply]
For getting the images' positions, `page.get_image_rects(xref)` does the job perfectly. Thanks a lot!

Regarding extracting the images, my question is about vector graphics. As far as I understand these objects are neither returned by `get_text()` nor they have xref. So both options wouldn't work. Is this correct? So what I need is to convert these vector graphics to an image/ images which can for example be displayed by `PIL.Image.open()`.

Regarding the difference in the outputs: The issue is that in the output of `page.get_text("dict")` some images are missing which are returned both by `get_images() `and `page.get_text("html")`. Those images are displayed on the same page I am processing.

**Reply by JorjMcKie** [reply]
You can extract a page's vector graphics via `page.get_drawings()`. This is low-level information about evry line, rectangle or curve is drawon. These atomic commands are grouped in a dictionary (pls see docu) which in PDF is called a "path", and hence in PyMuPDF too.
To make a nice picture, multiple paths are normally drawn.
To find out whether paths belong to each other, you can use a little [utility](https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/examples/extract-vector-graphics/separate-figures.py).

> Regarding the difference in the outputs: The issue is that in the output of page.get_text("dict") some images are missing which are returned both by get_images() and page.get_text("html"). Those images are displayed on the same page I am processing.

As I wrote: I would need to have that page.

**Reply by mahdyshabeeb** [reply]
I am really sorry for the late reply. I have sent you an email with the file where I am facing this issue.

**Reply by JorjMcKie** [reply]
@mahdyshabeeb - In my above list for possible explanations I forgot to mention this one:

The list `page.get_images()` **_does not contain the images displayed by the page!_**
Not in the general case that is.
This list simply quotes the image **_references_** found in the page's object definition. Not more, not less.
Therefore, if - for whatever reason - the PDF creator has decided to list no, or completely other images here, then this is legal and will not necessarily lead to errors.

Only `page.get_image_info()` and also `page.get_text("dict")` (they are both based on the same analysis of the page's appearance source) always reflect correctly the image display commands of the page.

If you execute `page.clean_contents()`, then an analysis of the page's appearance source is made and the page's object definition is synchronized accordingly with it.

**Reply by mahdyshabeeb** [reply]
Thanks for the explanation! The problem I'm facing is the other way around though. Using `page.get_text("dict")` some images are missing, which can be seen using  `page.get_images()` (and which are seen on the same page in the PDF file itself). So are these also somehow only referenced by the pdf creator but not really part of the page?

**Reply by JorjMcKie** [reply]
@mahdyshabeeb Cannot be!

The image list produced by `page.get_text("dict")` and `page.get_image_info()` is complete.
Try to prove the opposite 😎!
Your example file is no proof: after executing `page.clean_contents()`, `page.get_images()` coincides with that list.

**Reply by mahdyshabeeb** [reply]
I have tried executing `page.clean_contents()` and then `page.get_images()` but the results were exactly the same (as without using `page.clean_contents()`). I am still getting the images with xref 98 and 89, which I cannot see when running `page.get_text("dict")`. Since these images exist also in the pdf file, my question is, why are they not displayed by `page.get_text("dict")`? Is this the expected behaviour (should I always. use `page.get_text("dict")` and `page.get_images()` in combination)?
On the other hand, the method `page.get_image_info()` neither returns the binary representation of the images, nor the xref. So I'm not sure how to get the images when using that method.

**Reply by JorjMcKie** [reply]
You are talking of the file you sent me?
I tried that one too and confirmed that everything is fine. If you do this in this sequence:
1. page.clean_contents()
2. xrefs1 = [item[0] for item in page.get_images()]
3. xrefs2 = [item["xref"] for item in page.get_image_info(xrefs=True) if item["xref"] > 0]

Then you will see that both lists, xref1, xref2 are equal (at list as `set()`).

**Reply by JorjMcKie** [reply]
```python
page=doc[6]
page.clean_contents()
from pprint import pprint
pprint(page.get_images())
[(98, 0, 413, 578, 8, 'ICCBased', '', 'Im0', 'DCTDecode'),
 (100, 0, 122, 122, 8, 'ICCBased', '', 'Im1', 'DCTDecode'),
 (104, 103, 54, 66, 8, 'ICCBased', '', 'Im2', 'DCTDecode'),
 (89, 0, 411, 581, 8, 'ICCBased', '', 'Im0', 'DCTDecode')]
[i["xref"] for i in page.get_image_info(xrefs=True) if i["xref"]>0]
[98, 89, 100, 104]
```

**Reply by mahdyshabeeb** [reply]
Yes I am talking about that file.
So you would suggest using `page.get_image_info()` instead of `page.get_text("dict")`? Again my problem is that `page.get_text("dict")` does not return all the images. And I am trying to use this specific method since it returns both the text blocks and the images (and they can even be ordered) which is very useful for my use case.

**Reply by JorjMcKie** [reply]
Ah, I see your problem finally:
There is a difference between `page.get_text("dict")` and `page.get_image_info()`:
In text extraction, we are strictly observing the rectangle from which we exclusively extract if something is **_fully contained_** in the (explicit or implicit) clip rectangle.
This restriction does not exist with `page.get_image_info()`: we report all images anywhere on the page - even if parts are outside.
So if you do this: `clip = fitz.mediabox (-100, -100, 100, 100)` (i.e. enlarge the clip rectangle ficticiously by 100 in every direction), and then `page.get_text("dict", clip=clip)` then all images will be included.

**Reply by benmagos** [reply]
A-ha! Thanks for this last bit of helpful info.

I came here with the same discrepancy from images that are returned from `get_text("dict")` vs `page.get_image_info()`

In my case, I know that some images in question are far more than 100 outside of the MediaBox.
Is there any reasonable (perhaps from PDF Spec) limit to how far outside it can be?
e.g.:
```
clip = fitz.mediabox(-1000, -1000, 1000, 1000)
page.get_text("dict"), clip=clip)
```

For my needs, I would love to only use the results from `get_text("dict")` for both text/images, rather than using one for text and another for images, but understand if its currently simply more reliable to use `page.get_image_info()` if all images are needed.

**Reply by JorjMcKie** [reply]
Yeah, it is like that. The safe side is to take ~the mediabox and extend it by its own dimensions: `mb = page.mediabox`, then set clip to `clip = page.mediabox +(-mb.width, -mb.height, mb.width, mb.height)`~ `page.get_text("dict", clip=fitz.INFINITE_RECT())["blocks"]`.
Of course there is the risk to now getting unintended text too.

We had an internal discussion whether to include images in the clip restriction at all, or loosen the image selection to something like "include image if it intersects" etc.

But the decision in the end (and my favorite solution) was to use the same criteria for image and text.
If people like you are eccentric enough to absolutely see all images (🤣), then there also are ways to achieve that.

**Reply by JorjMcKie** [reply]
Sorry, should have thought of the infinite rectangle in the beginning.

**Reply by benmagos** [reply]
Many thanks for the context!

**Reply by mahdyshabeeb** [reply]
Thanks a lot Jorj! This is exactly what I was looking for.

**Reply by maxjeblick** [reply]
So, `page.get_image_info()` is roughly equivalent to (ignoring potential different return types) 
```
blocks = page.get_text("dict", sort=True, clip=fitz.INFINITE_RECT())["blocks"]
image_blocks = [block for block in blocks if block["type"] == 1]
```
?

**Reply by JorjMcKie** [reply]
@maxjeblick Correct! Another difference is that `get_image_info` does not return the image binaries ...

**Comment by enlacroix** [comment]
How to extract inline images via page.get_image_info() or page dict?

**Reply by JorjMcKie** [reply]
> How to extract inline images via page.get_image_info() or page dict?

This **_only_** works via `page.get_image_info()` because that method inspects the display instructions (the `/Contents`) of the page.
The Page dictionary is informal only in comparison. It only enumerates images that have an xref (inliners don't). Also, there is no guarantee that all enumerated images here are really displayed by the page etc.
Plus, there may be images **_with_** having an xref that are displayed, but are **_not_** in the page object definition. Example: images in annotations.

**Reply by enlacroix** [reply]
Thank you for such quick answer. I read all discussion and understood about inline specific images.
But, how to save inline images, that's my question. 
For images with xref I can write:
```python
  image_list = page.get_images()
  
  for image_index, img in enumerate(image_list):
      xref = img[0]
      position = page.get_image_info()[image_index - 1]["bbox"]
      base_image = pdf_file.extract_image(xref)
      image_bytes = base_image["image"]
      image_ext = base_image["ext"]
      extracted_image = Image.open(io.BytesIO(image_bytes)).convert('RGB')
```
But I see that not all images were extracted. If I'll use get_images_info I see that this method computes correct number of images on page. 
To sum up, how to save inline images?

**Reply by JorjMcKie** [reply]
All image items of a page are also contained in `page.get_text("dict")` and have `item["image"]` and `item["ext"]`. So you can do `pathlib.Path(f"my_image.{item['ext']}").write_bytes(item["image"])`.
This should save the image as a file with the appropriate extension.

Sorry - I gave the wrong info about the get_image_info(): this method only provides meta-information, not the image binary itself.

**Reply by JorjMcKie** [reply]
If an image has transparency (inliners don't), then there currently is no way to save the image as a transparent image, because we currently haven't implemented access to this so-called "mask" information.
